chain/tsresolver/tipset_resolver.go

```diff
@@ -0,0 +1,164 @@
+}
+
+var _ TipSetResolver = (*tipSetResolver)(nil)
+
+func NewTipSetResolver(loader TipSetLoader, f3 F3) TipSetResolver {
+	return &tipSetResolver{
+		loader: loader,
+		f3:     f3,
+	}
+}
+
+// ResolveEthBlockSelector resolves an Ethereum block selector string to a TipSet.
+//
+// The selector can be one of:
+//   - "pending": the chain head
+//   - "latest": the TipSet with the latest executed messages (head - 1)
+//   - "safe": the TipSet with messages executed at least eth.SafeEpochDelay (30) epochs ago or the
+//     latest F3 finalized TipSet
+//   - "finalized": the TipSet with messages executed at least policy.ChainFinality (900) epochs ago
+//     or the latest F3 finalized TipSet
+//   - a decimal block number: the TipSet at the given height
+//   - a 0x-prefixed hex block number: the TipSet at the given height
+//
+// If a specific block number is specified and `strict` is true, an error is returned if the block
+// number resolves to a null round, otherwise in the case of a null round the first non-null TipSet
+// immediately before the null round is returned.
+func (tsr *tipSetResolver) ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
+	switch selector {
+	case EthBlockSelectorEarliest:
+		return nil, fmt.Errorf(`block param "%s" is not supported`, EthBlockSelectorEarliest)
+
+	case EthBlockSelectorPending:
+		return tsr.loader.GetHeaviestTipSet(), nil
+
+	case EthBlockSelectorLatest:
+		// head - 1 because we're always one behind for Eth compatibility due to deferred execution
+		return tsr.loader.LoadTipSet(ctx, tsr.loader.GetHeaviestTipSet().Parents())
+
+	case EthBlockSelectorSafe, EthBlockSelectorFinalized:
+		defaultDelay := policy.ChainFinality
+		if selector == EthBlockSelectorSafe {
+			defaultDelay = ethtypes.SafeEpochDelay
+		}
+
+		head := tsr.loader.GetHeaviestTipSet()
+		latestHeight := head.Height() - 1 // always one behind for Eth compatibility due to deferred execution
+		defaultHeight := latestHeight - defaultDelay
+
+		if f3TipSet, err := tsr.getF3FinalizedTipSet(ctx); err != nil {
+			return nil, err
+		} else if f3TipSet != nil && f3TipSet.Height() > defaultHeight {
+			// return the parent of the finalized tipset (deferred execution, eth cares about t)
+			return tsr.loader.LoadTipSet(ctx, f3TipSet.Parents())
+		} // else F3 is disabled, not running, or behind the default safe or finalized height
+
+		ts, err := tsr.loader.GetTipsetByHeight(ctx, defaultHeight, head, true)
+		if err != nil {
+			return nil, xerrors.Errorf("loading tipset at height %v: %w", defaultHeight, err)
+		}
+		return ts, nil
+
+	default:
+		// likely an 0x hex block number or a decimal block number
+		return tsr.resolveEthBlockNumberSelector(ctx, selector, strict)
+	}
+}
+
+func (tsr *tipSetResolver) getF3FinalizedTipSet(ctx context.Context) (*types.TipSet, error) {
+	if !tsr.f3.IsEnabled() {
+		return nil, nil
+	}
+	if running, _ := tsr.f3.IsRunning(); !running {
+		return nil, nil
+	}
```

Comment:
We should have an additional check here for "is F3 running with finalization/static manifest".

Reply:
@rvagg, here is my take on this comment:

> is F3 running with finalization

This one means if "finalization" is enabled in F3 manifest, which in Lotus lingo translates to "Should F3 checkpoint the finalized tipsets?". It is specified [here](https://github.com/filecoin-project/go-f3/blob/d4f080decf5cd19c037b736263aedbf589faf8bf/manifest/manifest.go#L151-L152).

If finalisation in disabled in F3, then the API should be using EC finality.

> /static manifest

The F3 manifest integration in lotus is dynamic by default but up to a point, beyond which it will revert exclusively to a static manifest that "activates" F3 as the consensus mechanism for the network with uniform parameters across all nodes and no longer controllable by the F3 engineering team.

In the static manifest finalisation is enabled. If we are in a state where F3 is running, and we are in static manifest (i.e. F3 is active) then use F3 to resolve a tipset's finality.


@Kubuxu please correct me if I have missed anything.

Reply:
OK, thanks! I the static/dynamic thing has more nuance than I was aware of, I probably should go read the spec to understand that a bit more deeply.

Reply:
OK, went digging to understand a bit more of the dynamic/static situation and ended up having to dig into code. From what I can tell, there's no clear way of telling that we're not operating under a dynamic manifest, so the only check we can do here (the only check we need?) is to ask whether `manifest.EC.Finalize` is `true`, and if it's not, then ignore F3.

Does that sound right?

Reply:
I believe this is all handled now but I wouldn't mind a sanity check from both of you (@masih & @Kubuxu) in this function in particular.

It's now called `maybeF3FinalizedTipSet` to indicate that you may not get something useful from it. If F3 can't help us then we just move on to the defaults.

1. We're not going to error here unless it's from dealing with the chain store, an error from F3 is worthy of a WARN log, and only if the error isn't `ErrF3Disabled`.
2. If `!manifest.EC.Finalize`, we don't bother continuing
3. Some basic defensive checks on the certificate
4. All good? load and return that tipset, it's the latest that F3 has to offer.

Reply:
no, sadly this doesn't work:

```
runtime error: invalid memory address or nil pointer dereference
github.com/filecoin-project/go-jsonrpc.doCall.func1
	/home/runner/go/pkg/mod/github.com/filecoin-project/go-jsonrpc@v0.7.0/handler.go:237
runtime.gopanic
	/home/runner/work/_tool/go/1.22.7/x64/src/runtime/panic.go:770
runtime.panicmem
	/home/runner/work/_tool/go/1.22.7/x64/src/runtime/panic.go:261
runtime.sigpanic
	/home/runner/work/_tool/go/1.22.7/x64/src/runtime/signal_unix.go:881
github.com/filecoin-project/lotus/chain/lf3.(*F3).GetManifest
	/home/runner/work/lotus/lotus/chain/lf3/f3.go:205
github.com/filecoin-project/lotus/chain/tsresolver.(*tipSetResolver).maybeF3FinalizedTipSet
	/home/runner/work/lotus/lotus/chain/tsresolver/tipset_resolver.go:131
```

So `GetManifest` as it exists today isn't safe to call when it hasn't got a manifest. This is in an itest, letting it load the default `F3` module. I think maybe it needs to have a check in here for `nil`, but what should it return in that case? https://github.com/filecoin-project/lotus/blob/a46832f330215e047302d607753a59299a51eb45/chain/lf3/f3.go#L203-L206

Reply:
So we're querying it before the manifest provider can give it a manifest, I think `F3#GetManifest` should either return `nil, nil` or an error I could `errors.Is` on to deal with this case.

Reply:
Yeah, it initialises with its `manifest` `atomic.Pointer` empty, and it only gets set after `Start()` is called and `manifestProvider.ManifestUpdates()` successfully sends it a new manifest. https://github.com/filecoin-project/go-f3/blob/6c3d2deca27e361b68680156eef99271d82c29a6/f3.go#L170